packages/cozy-stack-client/src/OAuthClient.js

```diff
@@ -0,0 +1,358 @@
+import CozyStackClient from './CozyStackClient'
+import AccessToken from './AccessToken'
+
+const defaultoauthOptions = {
+  clientID: '',
+  clientName: '',
+  clientKind: '',
+  clientSecret: '',
+  clientURI: '',
+  registrationAccessToken: '',
+  redirectURI: '',
+  softwareID: '',
+  softwareVersion: '',
+  logoURI: '',
+  policyURI: '',
+  notificationPlatform: '',
+  notificationDeviceToken: ''
+}
+
+export default class OAuthClient extends CozyStackClient {
+  constructor({ oauth, scope, ...options }) {
+    super(options)
+    this.oauthOptions = { ...defaultoauthOptions, ...oauth }
+    this.scope = scope
+  }
+
+  /**
+   * Checks if the client has his registration informations from the server
+   * @returns {boolean}
+   * @private
+   */
+  isRegistered() {
+    return this.oauthOptions.clientID !== ''
+  }
+
+  /**
+   * Converts a camel-cased data set to snake case, suitable for sending to the OAuth server
+   * @param   {object} data Initial data
+   * @returns {object} Formatted data
+   * @private
+   */
+  snakeCaseOAuthData(data) {
+    const mappedFields = {
+      softwareID: 'software_id',
+      softwareVersion: 'software_version',
+      clientID: 'client_id',
+      clientName: 'client_name',
+      clientKind: 'client_kind',
+      clientURI: 'client_uri',
+      logoURI: 'logo_uri',
+      policyURI: 'policy_uri',
+      notificationPlatform: 'notification_platform',
+      notificationDeviceToken: 'notification_device_token',
+      redirectURI: 'redirect_uris'
+    }
+
+    const result = {}
+
+    Object.keys(data).forEach(fieldName => {
+      const key = mappedFields[fieldName] || fieldName
+      const value = data[fieldName]
+      result[key] = value
+    })
+
+    // special case: turn redirect_uris into an array
+    if (
+      result['redirect_uris'] &&
+      result['redirect_uris'] instanceof Array === false
+    )
+      result['redirect_uris'] = [result['redirect_uris']]
+
+    return result
+  }
```

Comment:
I am a bit impressed by this function. You put in inside a class not really related and the function's behavior own too much responsibilities: keep the full list of properties, build a binding object, transform a non array property into an array.

We could use regexp to transform property name, see an example here https://jsfiddle.net/LkmaLgsn/3/

And get the array transformation out of this function as it's not express by the name of the function.

Reply:
Maybe the functions are misnamed? What they really do is transform data in a format suitable for the server., and the other way around.

The regexp approach wouldn't work for names like `clientURI`, or the regexp would become quite complex, whereas the mapping here is very explicit.

Reply:
Here are my thoughts:
- this function is too big.
- this function has too much responsibilities.
- this function should not be public.

The regexp proposal was a proposition to help, I have no clue if it works or not, but it should as this function just transform a string into another string.

I can imagine something like a usage of this 3 one-lined functions:
```js
const toSnakeCase = prop =>
  prop.replace(/([A-Z]+)/g, "_$1").toLowerCase();

const redirect_uri_correction = prop =>
  /redirect_uri/.test(prop) ? `${prop}s` : prop;

const toArray = val => val instanceof Array ? val : [val]
```